components/style/animation.rs

```diff
@@ -369,70 +363,168 @@ impl PropertyAnimation {
-    context: &SharedStyleContext,
-    new_animations_sender: &Sender<Animation>,
-    opaque_node: OpaqueNode,
-    old_style: &ComputedValues,
-    new_style: &mut Arc<ComputedValues>,
-    expired_transitions: &[PropertyAnimation],
-) {
-    let mut all_running_animations = context.running_animations.write();
-    let previously_running_animations = all_running_animations
-        .remove(&opaque_node)
-        .unwrap_or_else(Vec::new);
-
-    let properties_that_transition = start_transitions_if_applicable(
-        context,
-        new_animations_sender,
-        opaque_node,
-        old_style,
-        new_style,
-        expired_transitions,
-        &previously_running_animations,
-    );
-
-    let mut all_cancelled_animations = context.cancelled_animations.write();
-    let mut cancelled_animations = all_cancelled_animations
-        .remove(&opaque_node)
-        .unwrap_or_else(Vec::new);
-    let mut running_animations = vec![];
-
-    // For every animation that was running before this style change, we cancel it
-    // if the property no longer transitions.
-    for running_animation in previously_running_animations.into_iter() {
-        if let Animation::Transition(_, _, ref property_animation) = running_animation {
-            if !properties_that_transition.contains(property_animation.property_id()) {
-                cancelled_animations.push(running_animation);
-                continue;
+/// Holds the animation state for a particular element.
+#[derive(Default)]
+pub struct ElementAnimationState {
+    /// The animations running for this element.
+    pub running_animations: Vec<Animation>,
+
+    /// The animations that have finished for this element, but not canceled. These are cleared
+    /// upon triggering a DOM event for each finished animation.
+    pub finished_animations: Vec<Animation>,
+
+    /// The animations that have been cancelled for this element. These are cleared
+    /// upon triggering a DOM event for each cancelled animation.
+    pub cancelled_animations: Vec<Animation>,
+
+    /// New animations created for this element.
+    pub new_animations: Vec<Animation>,
+}
+
+impl ElementAnimationState {
+    /// Cancel all animations in this `ElementAnimationState`. This is typically called
+    /// when the element has been removed from the DOM.
+    pub fn cancel_all_animations(&mut self) {
+        self.cancelled_animations.extend(
+            self.finished_animations
+                .drain(..)
+                .chain(self.running_animations.drain(..))
+                .chain(self.new_animations.drain(..)),
+        );
+    }
+
+    pub(crate) fn cancel_transitions_with_nontransitioning_properties(
+        &mut self,
+        properties_that_transition: LonghandIdSet,
+    ) {
+        if self.running_animations.is_empty() {
+            return;
+        }
+
+        let previously_running_transitions =
+            std::mem::replace(&mut self.running_animations, Vec::new());
```

Comment:
Maybe this could be marginally more efficient using `retain()` (you might need to replace the retained values with a dummy animation so you can move them to `self.cancelled_animations`, so maybe not a big deal.

Reply:
Is the issue here using a push for every animation. Another approach would be to rewrite `retain`, but instead of [truncating the vector](https://doc.rust-lang.org/src/alloc/vec.rs.html#1077-1097) it returns the partition index. Additionally, perhaps I could rely on `partition_in_place` though that isn't guaranteed to maintain the order. My thought was to figure out how to optimize these animation collections later, because maybe a `LinkedList` makes sense.